File: Java/프로그래머스/3/132266. 부대복귀/부대복귀.java
import java.util.*;
class Solution {
    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        int[] answer = new int[sources.length];
        Map<Integer, List<Integer>> map = new HashMap<>();
        for (int i = 1; i <= n; i++) {
            map.put(i, new ArrayList<>());
        }
        for (int[] road : roads) {
            map.get(road[0]).add(road[1]);
            map.get(road[1]).add(road[0]);
        }

        boolean[] visited = new boolean[n+1];
        int[] time = new int[n+1];
        Arrays.fill(time, -1);
        Deque<Integer> queue = new ArrayDeque<>();
        queue.offer(destination);
        visited[destination] = true;
        time[destination] = 0;

        while(!queue.isEmpty()) {
            int current = queue.poll();

            for (int next : map.get(current)) {
                if (!visited[next]) {
                    visited[next] = true;
                    queue.offer(next);
                    time[next] = time[current] + 1;
                }
            }
        }
        for (int i = 0; i < sources.length; i++) {
            answer[i] = time[sources[i]];
        }

        return answer;
    }
}

File: Python/프로그래머스/3/132266. 부대복귀/부대복귀.py
from collections import deque


def solution(n, roads, sources, destination):
    answer = [0] * len(sources)
    d = {}

    for i in range(1, n + 1):
        d[i] = []

    for a, b in roads:
        d.get(a).append(b)
        d.get(b).append(a)

    visited = [False] * (n + 1)
    time = [-1] * (n + 1)
    q = deque()
    q.append(destination)
    visited[destination] = True
    time[destination] = 0

    while q:
        cur = q.popleft()

        for next in d.get(cur):
            if not visited[next]:
                visited[next] = True
                q.append(next)
                time[next] = time[cur] + 1

    for i in range(len(sources)):
        answer[i] = time[sources[i]]

    return answer